### PR TITLE
Heatmap

### PR DESCRIPTION
**Pull request by mpodlasin**

### CODE REVIEW

File: src/client/visualizations/heat-map/heat-map.tsx [hunk 1]
```diff
@@ -16,7 +16,7 @@
  */
 
 import { TooltipWithBounds, withTooltip, WithTooltipProps } from "@vx/tooltip";
-import memoize = require("lodash.memoize");
```
Comment (adrianmroz) [inline src/client/visualizations/heat-map/heat-map.tsx]: remove `loads.memoize` from deps?

File: src/client/utils/dataset/dataset.mocha.ts [hunk 2]
```diff
@@ -0,0 +1,569 @@
...
+ */
+
+import { expect } from "chai";
+import { Timezone } from "chronoshift";
+import { Dataset } from "plywood";
+import { SPLIT } from "../../config/constants";
+import "../../utils/test-utils";
+import { fillDataset, orderByTimeDimensionDecreasing, orderByTimeDimensionIncreasing, orderByValueDecreasing, orderByValueIncreasing } from "./dataset";
+
+const rawDataset = {
```
Comment (adrianmroz) [inline src/client/utils/dataset/dataset.mocha.ts]: move to fixtures/snapshot files

File: src/client/utils/dataset/dataset.ts [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Timezone } from "chronoshift";
+import { Dataset, Datum, TimeRange } from "plywood";
+import { formatValue } from "../../../common/utils/formatter/formatter";
```
Comment (adrianmroz) [inline src/client/utils/dataset/dataset.ts]: formatting value when manipulating data seems wrong

File: src/client/utils/dataset/dataset.ts [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+};
+
+export const orderByValueIncreasing: Order<any> = (a, b) => {
+  return -orderByValueDecreasing(a, b);
+};
+
+export const orderByTimeDimensionDecreasing: Order<TimeRange> = ([_, __, originalA], [___, ____, originalB]) => originalA.compare(originalB);
+export const orderByTimeDimensionIncreasing: Order<TimeRange> = ([_, __, originalA], [___, ____, originalB]) => -originalA.compare(originalB);
+
+export const fillDataset = (dataset: Dataset, measureName: string, secondSplitName: string, order: Order<any>, timezone: Timezone): Dataset => {
```
Comment (adrianmroz) [inline src/client/utils/dataset/dataset.ts]: function and file names seems very generic

btw. do we need to return plywood dataset? POJsO wouldn't suffice?

Comment (mpodlasin) [reply src/client/utils/dataset/dataset.ts]: Using Dataset turned out a bit easier for testing - Dataset.toJS puts additional properties on objects - I would have to copy them all manually to my fixtures.

Comment (mpodlasin) [reply src/client/utils/dataset/dataset.ts]: Changed the name btw

File: src/client/visualizations/heat-map/heat-map.tsx [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+import { BaseVisualization, BaseVisualizationState } from "../base-visualization/base-visualization";
+import "./heat-map.scss";
+import { RectangleData } from "./heatmap-rectangles";
+import { LabelledHeatmap } from "./labelled-heatmap";
+
+const splitToFillSort = (split: Split): Sort<any> => {
+  const sort = split.sort;
+  switch (split.type) {
+    case SplitType.string:
+    case SplitType.number:
```
[the remark this answers is not in the document]
Comment (mpodlasin) [reply src/client/visualizations/heat-map/heat-map.tsx]: I tested. It seems to work well with number ranges.

File: src/client/visualizations/base-visualization/base-visualization.tsx [hunk 6]
```diff
@@ -37,10 +37,10 @@ export interface BaseVisualizationState {
   hoverMeasure?: Measure;
 }
 
-export class BaseVisualization<S extends BaseVisualizationState> extends React.Component<VisualizationProps, S> {
+export class BaseVisualization<S extends BaseVisualizationState, P = {}> extends React.Component<VisualizationProps & P, S> {
```
[the remark this answers is not in the document]
Comment (mpodlasin) [reply src/client/visualizations/base-visualization/base-visualization.tsx]: Well, I did what I said. Check it out!

File: src/client/visualizations/heat-map/heat-map.tsx [hunk 7]
```diff
@@ -98,13 +98,13 @@ export class UndecoratedHeatmapWithTooltip extends React.Component<Visualization
...
-    const { timezone, series } = essence;
-    const measure = essence.getEffectiveSelectedMeasures().first();
+    const { timezone } = essence;
+    const concreteSeries = essence.getConcreteSeries().first();
     const secondSplit = essence.splits.splits.get(1);
 
     const preparedDataset = memoizedFillDatasetWithMissingValues(
       (dataset.data[0][SPLIT] as Dataset),
-      measure.name,
+      concreteSeries.measure.name,
```
Comment (adrianmroz) [inline src/client/visualizations/heat-map/heat-map.tsx]: `plywoodKey`?

File: src/client/visualizations/heat-map/labelled-heatmap.tsx [hunk 8]
```diff
@@ -59,18 +60,16 @@ export class LabelledHeatmap extends React.PureComponent<Props, State> {
     const { dataset } = this.props;
     const { hoveredRectangle } = this.state;
 
-    const [measure] = this.props.essence.getEffectiveSelectedMeasures().toArray();
-    const [firstSplit, secondSplit] = this.props.essence.splits.splits.toArray();
+    const measure = this.props.essence.getEffectiveSelectedMeasures().first();
+    const [firstSplit, secondSplit] = this.props.essence.splits.splits.slice(0, 2).toArray();
```
Comment (adrianmroz) [inline src/client/visualizations/heat-map/labelled-heatmap.tsx]: You insist on keeping this in one line? :)

File: src/client/visualizations/table/table.tsx [hunk 9]
```diff
@@ -52,7 +52,7 @@ const SPACE_LEFT = 10;
 const SPACE_RIGHT = 10;
 const HIGHLIGHT_BUBBLE_V_OFFSET = -4;
 
-function formatSegment(value: any, timezone: Timezone): string {
+export function formatSegment(value: any, timezone: Timezone): string {
```
Comment (adrianmroz) [inline src/client/visualizations/table/table.tsx]: if exported, move to formatter.ts in utils

File: src/client/visualizations/heat-map/labelled-heatmap.tsx (hunk 8 above)
Comment (mpodlasin) [reply src/client/visualizations/heat-map/labelled-heatmap.tsx]: But I want to be cool and use ES6 :(

File: src/client/visualizations/table/table.tsx (hunk 9 above)
Comment (mpodlasin) [reply src/client/visualizations/table/table.tsx]: But then what is segment?

Should I change the name to `alsoFormatButDifferentlyThanFormatValue`? :D

File: package.json [hunk 10]
```diff
@@ -57,6 +57,12 @@
     "lint:sass": "sass-lint -v"
   },
   "dependencies": {
+    "@vx/event": "0.0.182",
+    "@vx/group": "0.0.183",
+    "@vx/heatmap": "0.0.183",
+    "@vx/mock-data": "0.0.185",
```
Comment (adrianmroz) [inline package.json]: ✂️

File: package.json [hunk 11]
```diff
@@ -102,8 +110,10 @@
     "@types/helmet": "0.0.37",
     "@types/js-yaml": "3.10.1",
     "@types/jsdom": "2.0.28",
+    "@types/lodash.memoize": "4.1.6",
```
Comment (adrianmroz) [inline package.json]: ✂️

Comment (mpodlasin) [reply package.json]: Done

File: src/client/visualizations/heat-map/heat-map.tsx [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+ * limitations under the License.
+ */
+
+// For some reason tsc compiler does not see this file.
+// Remove when issue is identified.
+// tslint:disable-next-line: no-reference
+/// <reference path="../../index.d.ts" />
+
+import { TooltipProps, TooltipWithBounds, withTooltip } from "@vx/tooltip";
+import memoize = require("memoizee");
```
Comment (adrianmroz) [inline src/client/visualizations/heat-map/heat-map.tsx]: looks weird :D

Comment (mpodlasin) [reply src/client/visualizations/heat-map/heat-map.tsx]: proper TypeScript :D

Comment (mpodlasin) [reply src/client/visualizations/heat-map/heat-map.tsx]: Other imports didn't work

Comment (adrianmroz) [reply src/client/visualizations/heat-map/heat-map.tsx]: It's not critical for now, but we need to check why.

File: src/client/visualizations/heat-map/heat-map.tsx [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+      <LabelledHeatmap
+        dataset={preparedDataset}
+        essence={this.props.essence}
+        onHover={this.handleRectangleHover}
+        onHoverStop={hideTooltip}
+      />
+      {tooltipOpen && (
+        <TooltipWithBounds
+          // set this to random so it correctly updates with parent bounds
+          key={Math.random()}
```
Comment (adrianmroz) [inline src/client/visualizations/heat-map/heat-map.tsx]: 😵

File: src/client/visualizations/heat-map/heatmap-labels.tsx [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+  componentDidMount() {
+    if (this.container === null) {
+      return;
+    }
+
+    const {
+      onMaxLabelSize = () => {}
+    } = this.props;
+
+    const maxWidth = Array.from(this.container.querySelectorAll("span > span")).reduce((maxWidth, element: HTMLSpanElement) => {
```
Comment (adrianmroz) [inline src/client/visualizations/heat-map/heatmap-labels.tsx]: error prone selector

in css you also have few `> *`, why not give proper classes?

File: src/client/visualizations/heat-map/heatmap-labels.tsx [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+      if (element.offsetWidth > maxWidth) {
```
Comment (adrianmroz) [inline src/client/visualizations/heat-map/heatmap-labels.tsx]: `return Math.max(element.offsetWidth, maxWidth)`

on if less :D

File: src/client/visualizations/heat-map/heat-map.tsx (hunk 12 above)
Comment (mpodlasin) [reply src/client/visualizations/heat-map/heat-map.tsx]: https://www.typescriptlang.org/docs/handbook/modules.html - older packages for compatibility. It looks weird, but it's 100% ok to use.

File: src/common/manifests/heat-map/heat-map.ts [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+import { Split, SplitType } from "../../models/split/split";
+import { Predicates } from "../../utils/rules/predicates";
+import { visualizationDependentEvaluatorBuilder } from "../../utils/rules/visualization-dependent-evaluator";
+
+const rulesEvaluator = visualizationDependentEvaluatorBuilder
+  .when(Predicates.numberOfSplitsIsNot(2))
+  .then(({ dataCube, splits }) => Resolve.manual(
+    3,
+    "Heatmap needs exactly 2 splits",
+    splits.length() > 2 ?
```
Comment (adrianmroz) [inline src/common/manifests/heat-map/heat-map.ts]: The ternary reads really badly ...

File: src/common/utils/rules/predicates.ts [hunk 17]
```diff
@@ -24,6 +24,14 @@ export class Predicates {
     return ({ splits }) => splits.length() === 0;
   }
 
+  public static numberOfSplitsIsNot(expected: number): VisualizationDependentPredicate {
+    return ({ splits }) => splits.length() !== expected;
+  }
+
+  public static numberOfMeasuresIsNot(expected: number): VisualizationDependentPredicate {
```
Comment (adrianmroz) [inline src/common/utils/rules/predicates.ts]: `numberOfSeries`

File: src/common/models/dimension/dimensions.ts [hunk 18]
```diff
@@ -132,4 +132,8 @@ export class Dimensions {
   toJS(): DimensionOrGroupJS[] {
     return this.dimensions.map(dimensionOrGroup => dimensionOrGroup.toJS());
   }
+
+  slice(begin: number, end?: number): Dimensions {
```
Comment (adrianmroz) [inline src/common/models/dimension/dimensions.ts]: Do we need `Dimensions` class? Constructor looks costly, maybe we could just slice `flattenedDimensions` and return `Dimension[]`?

File: src/client/visualizations/heat-map/heat-map.tsx (hunk 13 above)
Comment (mpodlasin) [reply src/client/visualizations/heat-map/heat-map.tsx]: It's from vx docs. It really does not work properly without it 😂

File: src/client/visualizations/heat-map/heatmap-labels.tsx (hunk 14 above)
Comment (mpodlasin) [reply src/client/visualizations/heat-map/heatmap-labels.tsx]: Done

File: src/client/visualizations/heat-map/heatmap-labels.tsx (hunk 15 above)
Comment (mpodlasin) [reply src/client/visualizations/heat-map/heatmap-labels.tsx]: Let me overcomplicate my code! ;)

Done, of course

File: src/common/manifests/heat-map/heat-map.ts (hunk 16 above)
Comment (mpodlasin) [reply src/common/manifests/heat-map/heat-map.ts]: Refactored

File: src/common/utils/rules/predicates.ts (hunk 17 above)
Comment (mpodlasin) [reply src/common/utils/rules/predicates.ts]: Done

File: src/common/models/dimension/dimensions.ts (hunk 18 above)
Comment (mpodlasin) [reply src/common/models/dimension/dimensions.ts]: It was not used at all 😂

File: src/client/visualizations/heat-map/heatmap-rectangles.tsx [hunk 19]
```diff
@@ -0,0 +1,192 @@
...
+      xLabel: hoveredBins[leftLabelName] as string,
+      yLabel: hoveredBin[topLabelName] as string,
+      x,
+      y,
+      column: yPosition,
+      row: xPosition
+    });
+  }
+
+  private setup() {
```
Comment (adrianmroz) [inline src/client/visualizations/heat-map/heatmap-rectangles.tsx]: We could win some performance with smart memoization of this code. Each render we're running through whole dataset and thrash scales...

**Pull request merged**